File: android-katha-ai/app/src/main/java/com/rork/kathaai/data/AnalyticsService.kt
package com.rork.kathaai.data

import android.content.Context
import com.rork.kathaai.model.*
import kotlinx.serialization.json.Json
import java.util.Calendar
import java.util.UUID

object AnalyticsService {
    private const val KEY_ANALYTICS = "katha.storyAnalytics"
    private const val KEY_UNSEEN = "katha.unseenMilestones"
    private const val KEY_AUTHOR_MILESTONES = "katha.authorMilestonesCrossed"
    private const val KEY_TIME_RANGE = "katha.analyticsTimeRange"

    private val json = Json { ignoreUnknownKeys = true }

    private var storyAnalytics: MutableMap<String, StoryAnalytics> = mutableMapOf()
    private var unseenMilestones: MutableList<UnseenMilestone> = mutableListOf()
    private var authorMilestonesCrossed: MutableSet<String> = mutableSetOf()
    private var seeded = false
    private var prefs: android.content.SharedPreferences? = null

    fun init(context: Context) {
        if (prefs != null) return
        prefs = context.getSharedPreferences("katha", Context.MODE_PRIVATE)
        load()
    }

    private fun load() {
        val p = prefs ?: return
        p.getString(KEY_ANALYTICS, null)?.let { raw ->
            runCatching {
                val map = json.decodeFromString<Map<String, StoryAnalytics>>(raw)
                storyAnalytics = map.toMutableMap()
                seeded = true
            }
        }
        p.getString(KEY_UNSEEN, null)?.let { raw ->
            runCatching { unseenMilestones = json.decodeFromString(raw) }
        }
        authorMilestonesCrossed = (p.getStringSet(KEY_AUTHOR_MILESTONES, emptySet()) ?: emptySet()).toMutableSet()
    }

    private fun saveAnalytics() {
        val p = prefs ?: return
        p.edit().putString(KEY_ANALYTICS, json.encodeToString(storyAnalytics)).apply()
    }

    private fun saveUnseen() {
        val p = prefs ?: return
        p.edit().putString(KEY_UNSEEN, json.encodeToString(unseenMilestones)).apply()
    }

    private fun saveAuthorMilestones() {
        val p = prefs ?: return
        p.edit().putStringSet(KEY_AUTHOR_MILESTONES, authorMilestonesCrossed).apply()
    }

    // MARK: - Time Range

    fun getTimeRange(): AnalyticsTimeRange {
        val ordinal = prefs?.getInt(KEY_TIME_RANGE, 1) ?: 1
        return AnalyticsTimeRange.entries.getOrElse(ordinal) { AnalyticsTimeRange.THIRTY_DAYS }
    }

    fun setTimeRange(range: AnalyticsTimeRange) {
        prefs?.edit()?.putInt(KEY_TIME_RANGE, range.ordinal)?.apply()
    }

    // MARK: - Seed

    fun ensureSeeded(stories: List<Story>) {
        if (seeded) return
        seeded = true

        for (story in stories) {
            if (storyAnalytics[story.id] == null) {
                storyAnalytics[story.id] = generateMockAnalytics(story)
            }
        }

        // Aarav's story: 47 reads, 4 credits
        stories.firstOrNull { it.authorId == "aarav" }?.let { aaravStory ->
            storyAnalytics[aaravStory.id] = generateAaravAnalytics(aaravStory)
        }

        saveAnalytics()
    }

    private fun generateMockAnalytics(story: Story): StoryAnalytics {
        val totalReads = story.views
        val dailyReads = generateDailyReads(totalReads, story.publishedOffset)
        val chapterReads = generateChapterReads(story, totalReads)
        val creditsEarned = totalReads / 12
        val storyFollowers = story.followerCount
        val authorFollowers = SeedData.author(story.authorId)?.followers ?: 0

        val crossed = computeMilestones(totalReads, storyFollowers, creditsEarned, true, authorFollowers)

        return StoryAnalytics(
            dailyReads = dailyReads,
            chapterReads = chapterReads,
            milestonesCrossed = crossed,
            creditsEarnedFromReads = creditsEarned,
            followersGainedFromStory = storyFollowers,
            authorFollowersGainedViaStory = maxOf(0, storyFollowers / 3)
        )
    }

    private fun generateAaravAnalytics(story: Story): StoryAnalytics {
        val totalReads = 47
        val dailyReads = generateDailyReads(totalReads, 30)
        val chapterReads = generateChapterReads(story, totalReads)

        val crossed = setOf(
            MilestoneKey.PUBLISH.key,
            MilestoneKey.FIRST_READ.key,
            MilestoneKey.READS_10.key,
            MilestoneKey.FIRST_CREDIT.key
        )

        return StoryAnalytics(
            dailyReads = dailyReads,
            chapterReads = chapterReads,
            milestonesCrossed = crossed,
            creditsEarnedFromReads = 4,
            followersGainedFromStory = 8,
            authorFollowersGainedViaStory = 12
        )
    }

    private fun generateDailyReads(totalReads: Int, publishedOffset: Int): List<DailyReads> {
        val cal = Calendar.getInstance()
        val now = System.currentTimeMillis()
        val days = minOf(maxOf(7, publishedOffset), 90)

        if (totalReads == 0) {
            return (0 until minOf(days, 7)).map { offset ->
                cal.timeInMillis = now - offset * 24L * 3600 * 1000
                DailyReads(dateISO = isoDate(cal), reads = 0, uniqueReads = 0)
            }.reversed()
        }

        val weights = DoubleArray(days) { i ->
            val recencyBias = 1.0 + i.toDouble() / days * 0.6
            cal.timeInMillis = now - (days - 1 - i) * 24L * 3600 * 1000
            val weekday = cal.get(Calendar.DAY_OF_WEEK)
            val weekendBoost = if (weekday == 1 || weekday == 7) 1.2 else 1.0
            val noise = 0.7 + Math.random() * 0.6
            recencyBias * weekendBoost * noise
        }
        val sum = weights.sum()
        val normalized = IntArray(days) { i ->
            maxOf(0, (weights[i] / sum * totalReads).toInt())
        }
        val diff = totalReads - normalized.sum()
        if (normalized.isNotEmpty()) normalized[normalized.size - 1] = maxOf(0, normalized[normalized.size - 1] + diff)

        return normalized.indices.map { i ->
            cal.timeInMillis = now - (days - 1 - i) * 24L * 3600 * 1000
            DailyReads(dateISO = isoDate(cal), reads = normalized[i], uniqueReads = (normalized[i] * 0.75).toInt())
        }
    }

    private fun generateChapterReads(story: Story, totalReads: Int): List<ChapterReads> {
        if (story.chapters.size <= 1) return emptyList()

        val chapterCount = story.chapters.size
        val weights = DoubleArray(chapterCount) { i -> Math.pow(0.72, i.toDouble()) }
        val sum = weights.sum()

        return story.chapters.mapIndexed { index, chapter ->
            val reads = maxOf(0, (totalReads * weights[index] / sum).toInt())
            ChapterReads(
                chapterId = chapter.id,
                reads = reads,
                uniqueReads = (reads * 0.75).toInt(),
                completionRate = maxOf(0.2, minOf(0.95, 1.0 - index * 0.15))
            )
        }
    }

    private fun computeMilestones(totalReads: Int, storyFollowers: Int, creditsEarned: Int, isPublished: Boolean, authorFollowers: Int): Set<String> {
        val crossed = mutableSetOf<String>()
        for (milestone in MilestoneKey.entries) {
            if (milestone.isCrossed(totalReads, storyFollowers, creditsEarned, isPublished, authorFollowers)) {
                crossed.add(milestone.key)
            }
        }
        return crossed
    }

    private fun isoDate(cal: Calendar): String {
        val y = cal.get(Calendar.YEAR)
        val m = cal.get(Calendar.MONTH) + 1
        val d = cal.get(Calendar.DAY_OF_MONTH)
        return "%04d-%02d-%02d".format(y, m, d)
    }

    // MARK: - Access

    fun analyticsFor(storyId: String): StoryAnalytics? = storyAnalytics[storyId]

    fun dailyReadsFor(storyId: String, range: AnalyticsTimeRange): List<DailyReads> {
        val analytics = storyAnalytics[storyId] ?: return emptyList()
        return when (range) {
            AnalyticsTimeRange.SEVEN_DAYS -> analytics.dailyReads.takeLast(7)
            AnalyticsTimeRange.THIRTY_DAYS -> analytics.dailyReads.takeLast(30)
            AnalyticsTimeRange.ALL_TIME -> analytics.dailyReads
        }
    }

    fun totalReadsFor(storyId: String, range: AnalyticsTimeRange): Int =
        dailyReadsFor(storyId, range).sumOf { it.reads }

    fun previousPeriodReads(storyId: String, range: AnalyticsTimeRange): Int {
        val analytics = storyAnalytics[storyId] ?: return 0
        val all = analytics.dailyReads
        val dayCount = range.dayCount ?: all.size
        val endIdx = maxOf(0, all.size - dayCount)
        val startIdx = maxOf(0, endIdx - dayCount)
        return all.subList(startIdx, endIdx).sumOf { it.reads }
    }

    // MARK: - Milestones

    fun hasUnseenMilestones(): Boolean = unseenMilestones.isNotEmpty()
    fun currentUnseenMilestone(): UnseenMilestone? = unseenMilestones.firstOrNull()

    fun dismissCurrentMilestone() {
        if (unseenMilestones.isNotEmpty()) {
            unseenMilestones.removeAt(0)
            saveUnseen()
        }
    }

    // MARK: - Dev Tools

    fun triggerMilestoneCelebration(milestone: MilestoneKey, storyId: String, storyTitle: String) {
        unseenMilestones.add(UnseenMilestone(
            id = UUID.randomUUID().toString(),
            milestoneKey = milestone.key,
            storyId = storyId,
            storyTitle = storyTitle
        ))
        saveUnseen()
    }

    fun resetAllMilestones() {
        for (key in storyAnalytics.keys) {
            storyAnalytics[key] = storyAnalytics[key]!!.copy(milestonesCrossed = emptySet())
        }
        authorMilestonesCrossed.clear()
        unseenMilestones.clear()
        saveAnalytics()
        saveAuthorMilestones()
        saveUnseen()
    }

    fun resetAllAnalytics() {
        storyAnalytics.clear()
        unseenMilestones.clear()
        authorMilestonesCrossed.clear()
        seeded = false
        prefs?.edit()?.remove(KEY_ANALYTICS)?.remove(KEY_UNSEEN)?.remove(KEY_AUTHOR_MILESTONES)?.apply()
    }

    // MARK: - Aggregated

    fun aggregatedDailyReads(storyIds: List<String>, range: AnalyticsTimeRange): List<DailyReads> {
        val merged = mutableMapOf<String, Int>()
        val orderedDates = mutableListOf<String>()

        for (storyId in storyIds) {
            for (entry in dailyReadsFor(storyId, range)) {
                if (merged[entry.dateISO] == null) orderedDates.add(entry.dateISO)
                merged[entry.dateISO] = (merged[entry.dateISO] ?: 0) + entry.reads
            }
        }

        return orderedDates.map { iso ->
            DailyReads(dateISO = iso, reads = merged[iso] ?: 0, uniqueReads = ((merged[iso] ?: 0) * 0.75).toInt())
        }
    }

    fun totalReadsAcrossCatalog(storyIds: List<String>, range: AnalyticsTimeRange): Int =
        storyIds.sumOf { totalReadsFor(it, range) }

    fun totalCreditsAcrossCatalog(storyIds: List<String>): Int =
        storyIds.sumOf { storyAnalytics[it]?.creditsEarnedFromReads ?: 0 }

    fun recentMilestones(limit: Int = 5): List<Triple<MilestoneKey, String, String>> {
        val results = mutableListOf<Triple<MilestoneKey, String, String>>()
        for ((storyId, analytics) in storyAnalytics) {
            for (key in analytics.milestonesCrossed) {
                val milestone = MilestoneKey.fromKey(key) ?: continue
                val story = SeedData.stories.firstOrNull { it.id == storyId }
                results.add(Triple(milestone, storyId, story?.title ?: ""))
            }
        }
        return results.take(limit)
    }

    fun topStories(storyIds: List<String>, sortBy: MilestoneSortMetric, limit: Int = 5): List<String> {
        return when (sortBy) {
            MilestoneSortMetric.READS -> storyIds.sortedByDescending { totalReadsFor(it, AnalyticsTimeRange.ALL_TIME) }
            MilestoneSortMetric.CREDITS -> storyIds.sortedByDescending { storyAnalytics[it]?.creditsEarnedFromReads ?: 0 }
            MilestoneSortMetric.FOLLOWERS -> storyIds.sortedByDescending { storyAnalytics[it]?.followersGainedFromStory ?: 0 }
        }.take(limit)
    }
}

File: ios-katha-ai-create-stories/KathaAICreateStories/Services/AnalyticsService.swift
//
//  AnalyticsService.swift
//  KathaAICreateStories
//

import Foundation

// MARK: - Analytics Service

final class AnalyticsService {
    static let shared = AnalyticsService()

    private let defaults = UserDefaults.standard
    private let analyticsKey = "katha.storyAnalytics"
    private let authorMilestonesKey = "katha.authorMilestonesCrossed"
    private let unseenMilestonesKey = "katha.unseenMilestones"
    private let timeRangeKey = "katha.analyticsTimeRange"

    // In-memory cache
    private(set) var storyAnalytics: [String: StoryAnalytics] = [:]
    private(set) var authorMilestonesCrossed: Set<String> = []
    private(set) var unseenMilestones: [UnseenMilestone] = []

    private var seeded = false

    private init() {
        load()
    }

    // MARK: - Load / Save

    private func load() {
        if let data = defaults.data(forKey: analyticsKey),
           let decoded = try? JSONDecoder().decode([String: StoryAnalytics].self, from: data) {
            storyAnalytics = decoded
            seeded = true
        }
        if let data = defaults.data(forKey: unseenMilestonesKey),
           let decoded = try? JSONDecoder().decode([UnseenMilestone].self, from: data) {
            unseenMilestones = decoded
        }
        authorMilestonesCrossed = Set(defaults.stringArray(forKey: authorMilestonesKey) ?? [])
    }

    private func saveAnalytics() {
        if let data = try? JSONEncoder().encode(storyAnalytics) {
            defaults.set(data, forKey: analyticsKey)
        }
    }

    private func saveUnseenMilestones() {
        if let data = try? JSONEncoder().encode(unseenMilestones) {
            defaults.set(data, forKey: unseenMilestonesKey)
        }
    }

    private func saveAuthorMilestones() {
        defaults.set(Array(authorMilestonesCrossed), forKey: authorMilestonesKey)
    }

    // MARK: - Time Range

    var timeRange: AnalyticsTimeRange {
        AnalyticsTimeRange(rawValue: defaults.integer(forKey: timeRangeKey)) ?? .thirtyDays
    }

    func setTimeRange(_ range: AnalyticsTimeRange) {
        defaults.set(range.rawValue, forKey: timeRangeKey)
    }

    // MARK: - Seed Mock Analytics

    func ensureSeeded(stories: [Story]) {
        guard !seeded else { return }
        seeded = true

        for story in stories {
            guard storyAnalytics[story.id] == nil else { continue }
            storyAnalytics[story.id] = generateMockAnalytics(for: story)
        }

        // Aarav's specific story: 47 reads, 4 credits, milestones
        if let aaravStory = stories.first(where: { $0.authorId == "aarav" }) {
            storyAnalytics[aaravStory.id] = generateAaravAnalytics(for: aaravStory)
        }

        saveAnalytics()
    }

    private func generateMockAnalytics(for story: Story) -> StoryAnalytics {
        let totalReads = story.views
        let dailyReads = generateDailyReads(totalReads: totalReads, publishedOffset: story.publishedOffset)
        let chapterReads = generateChapterReads(story: story, totalReads: totalReads)
        let creditsEarned = totalReads / 12
        let storyFollowers = story.followerCount

        let crossed = computeMilestones(
            totalReads: totalReads,
            storyFollowers: storyFollowers,
            creditsEarned: creditsEarned,
            isPublished: true,
            authorFollowers: SeedData.author(id: story.authorId)?.followers ?? 0
        )

        return StoryAnalytics(
            dailyReads: dailyReads,
            chapterReads: chapterReads,
            milestonesCrossed: crossed,
            creditsEarnedFromReads: creditsEarned,
            followersGainedFromStory: storyFollowers,
            authorFollowersGainedViaStory: max(0, storyFollowers / 3),
            lastUpdated: Date()
        )
    }

    private func generateAaravAnalytics(for story: Story) -> StoryAnalytics {
        let totalReads = 47
        let dailyReads = generateDailyReads(totalReads: totalReads, publishedOffset: 30)
        let chapterReads = generateChapterReads(story: story, totalReads: totalReads)
        let creditsEarned = 4

        let crossed: Set<String> = [
            MilestoneKey.publish.rawValue,
            MilestoneKey.firstRead.rawValue,
            MilestoneKey.reads10.rawValue,
            MilestoneKey.firstCredit.rawValue
        ]

        return StoryAnalytics(
            dailyReads: dailyReads,
            chapterReads: chapterReads,
            milestonesCrossed: crossed,
            creditsEarnedFromReads: creditsEarned,
            followersGainedFromStory: 8,
            authorFollowersGainedViaStory: 12,
            lastUpdated: Date()
        )
    }

    private func generateDailyReads(totalReads: Int, publishedOffset: Int) -> [DailyReads] {
        let cal = Calendar.current
        let now = Date()
        let days = min(max(7, publishedOffset), 90)

        guard totalReads > 0 else {
            // Generate zero entries for 7 days
            return (0..<min(days, 7)).map { offset in
                let date = cal.date(byAdding: .day, value: -offset, to: now)!
                let iso = AnalyticsDateFormatter.shared.string(from: date)
                return DailyReads(dateISO: iso, reads: 0, uniqueReads: 0)
            }.reversed()
        }

        var weights: [Double] = []
        for i in 0..<days {
            let recencyBias = 1.0 + Double(i) / Double(days) * 0.6
            let date = cal.date(byAdding: .day, value: -(days - 1 - i), to: now)!
            let weekday = cal.component(.weekday, from: date)
            let weekendBoost = (weekday == 1 || weekday == 7) ? 1.2 : 1.0
            let noise = 0.7 + Double.random(in: 0...0.6)
            weights.append(recencyBias * weekendBoost * noise)
        }
        let sum = weights.reduce(0, +)
        var normalized = weights.map { max(0, Int(round(Double($0) / sum * Double(totalReads)))) }
        let diff = totalReads - normalized.reduce(0, +)
        if !normalized.isEmpty { normalized[normalized.count - 1] = max(0, normalized[normalized.count - 1] + diff) }

        return normalized.enumerated().map { index, reads in
            let date = cal.date(byAdding: .day, value: -(days - 1 - index), to: now)!
            let iso = AnalyticsDateFormatter.shared.string(from: date)
            return DailyReads(dateISO: iso, reads: reads, uniqueReads: Int(Double(reads) * 0.75))
        }
    }

    private func generateChapterReads(story: Story, totalReads: Int) -> [ChapterReads] {
        guard story.chapters.count > 1 else { return [] }

        let chapterCount = story.chapters.count
        var weights: [Double] = []
        for i in 0..<chapterCount {
            let dropOff = pow(0.72, Double(i))
            weights.append(dropOff)
        }
        let sum = weights.reduce(0, +)
        return story.chapters.enumerated().map { index, chapter in
            let reads = max(0, Int(round(Double(totalReads) * weights[index] / sum)))
            return ChapterReads(
                chapterId: chapter.id,
                reads: reads,
                uniqueReads: Int(Double(reads) * 0.75),
                completionRate: max(0.2, min(0.95, 1.0 - Double(index) * 0.15))
            )
        }
    }

    private func computeMilestones(totalReads: Int, storyFollowers: Int, creditsEarned: Int, isPublished: Bool, authorFollowers: Int) -> Set<String> {
        var crossed: Set<String> = []
        for milestone in MilestoneKey.allCases {
            if milestone.isCrossed(
                totalReads: totalReads,
                storyFollowers: storyFollowers,
                creditsEarned: creditsEarned,
                isPublished: isPublished,
                authorFollowers: authorFollowers
            ) {
                crossed.insert(milestone.rawValue)
            }
        }
        return crossed
    }

    // MARK: - Analytics Access

    func analytics(for storyId: String) -> StoryAnalytics? {
        storyAnalytics[storyId]
    }

    func dailyReads(for storyId: String, range: AnalyticsTimeRange) -> [DailyReads] {
        guard let analytics = storyAnalytics[storyId] else { return [] }
        switch range {
        case .sevenDays:
            return Array(analytics.dailyReads.suffix(7))
        case .thirtyDays:
            return Array(analytics.dailyReads.suffix(30))
        case .allTime:
            return analytics.dailyReads
        }
    }

    func totalReads(for storyId: String, range: AnalyticsTimeRange) -> Int {
        dailyReads(for: storyId, range: range).reduce(0) { $0 + $1.reads }
    }

    func previousPeriodReads(for storyId: String, range: AnalyticsTimeRange) -> Int {
        guard let analytics = storyAnalytics[storyId] else { return 0 }
        let all = analytics.dailyReads
        let dayCount = range.dayCount ?? all.count
        let endIdx = max(0, all.count - dayCount)
        let startIdx = max(0, endIdx - dayCount)
        return all[startIdx..<endIdx].reduce(0) { $0 + $1.reads }
    }

    func creditsEarned(for storyId: String, range: AnalyticsTimeRange) -> Int {
        guard let analytics = storyAnalytics[storyId] else { return 0 }
        return analytics.creditsEarnedFromReads
    }

    // MARK: - Milestone Checking

    func checkMilestones(storyId: String, title: String, totalReads: Int, storyFollowers: Int, creditsEarned: Int, isPublished: Bool, authorFollowers: Int) {
        guard let analytics = storyAnalytics[storyId] else { return }

        let currentCrossed = computeMilestones(
            totalReads: totalReads,
            storyFollowers: storyFollowers,
            creditsEarned: creditsEarned,
            isPublished: isPublished,
            authorFollowers: authorFollowers
        )

        let newlyCrossed = currentCrossed.subtracting(analytics.milestonesCrossed)

        if !newlyCrossed.isEmpty {
            var updated = analytics
            updated.milestonesCrossed = currentCrossed
            updated.lastUpdated = Date()
            storyAnalytics[storyId] = updated
            saveAnalytics()

            for key in newlyCrossed {
                guard let milestone = MilestoneKey(rawValue: key) else { continue }
                let unseen = UnseenMilestone(
                    id: UUID().uuidString,
                    milestoneKey: key,
                    storyId: storyId,
                    storyTitle: title,
                    timestamp: Date()
                )
                unseenMilestones.append(unseen)

                // Also check author-level milestones
                if milestone == .authorFollowers100 {
                    authorMilestonesCrossed.insert(key)
                    saveAuthorMilestones()
                }
            }
            saveUnseenMilestones()
        }
    }

    // MARK: - Unseen Milestones Queue

    var currentUnseenMilestone: UnseenMilestone? {
        unseenMilestones.first
    }

    var hasUnseenMilestones: Bool {
        !unseenMilestones.isEmpty
    }

    func dismissCurrentMilestone() {
        guard !unseenMilestones.isEmpty else { return }
        unseenMilestones.removeFirst()
        saveUnseenMilestones()
    }

    // MARK: - Dev Tools

    func triggerMilestoneCelebration(milestone: MilestoneKey, storyId: String, storyTitle: String) {
        let unseen = UnseenMilestone(
            id: UUID().uuidString,
            milestoneKey: milestone.rawValue,
            storyId: storyId,
            storyTitle: storyTitle,
            timestamp: Date()
        )
        unseenMilestones.append(unseen)
        saveUnseenMilestones()
    }

    func resetAllMilestones() {
        for key in storyAnalytics.keys {
            storyAnalytics[key]?.milestonesCrossed = []
        }
        authorMilestonesCrossed = []
        unseenMilestones = []
        saveAnalytics()
        saveAuthorMilestones()
        saveUnseenMilestones()
    }

    func resetAllAnalytics() {
        storyAnalytics = [:]
        unseenMilestones = []
        authorMilestonesCrossed = []
        seeded = false
        defaults.removeObject(forKey: analyticsKey)
        defaults.removeObject(forKey: unseenMilestonesKey)
        defaults.removeObject(forKey: authorMilestonesKey)
    }

    // MARK: - Aggregated Dashboard Data

    func aggregatedDailyReads(storyIds: [String], range: AnalyticsTimeRange) -> [DailyReads] {
        var merged: [String: Int] = [:]
        var orderedDates: [String] = []

        for storyId in storyIds {
            let reads = dailyReads(for: storyId, range: range)
            for entry in reads {
                if merged[entry.dateISO] == nil { orderedDates.append(entry.dateISO) }
                merged[entry.dateISO, default: 0] += entry.reads
            }
        }

        return orderedDates.map { iso in
            DailyReads(dateISO: iso, reads: merged[iso] ?? 0, uniqueReads: Int(Double(merged[iso] ?? 0) * 0.75))
        }
    }

    func totalReadsAcrossCatalog(storyIds: [String], range: AnalyticsTimeRange) -> Int {
        storyIds.reduce(0) { $0 + totalReads(for: $1, range: range) }
    }

    func totalCreditsAcrossCatalog(storyIds: [String]) -> Int {
        storyIds.reduce(0) { $0 + (storyAnalytics[$1]?.creditsEarnedFromReads ?? 0) }
    }

    func recentMilestones(limit: Int = 5) -> [(milestone: MilestoneKey, storyId: String, storyTitle: String, timestamp: Date)] {
        var results: [(MilestoneKey, String, String, Date)] = []
        for (storyId, analytics) in storyAnalytics {
            for key in analytics.milestonesCrossed {
                guard let milestone = MilestoneKey(rawValue: key) else { continue }
                let story = SeedData.stories.first(where: { $0.id == storyId })
                results.append((milestone, storyId, story?.title ?? "", analytics.lastUpdated))
            }
        }
        return results.sorted { $0.3 > $1.3 }.prefix(limit).map { $0 }
    }

    func topStories(storyIds: [String], sortBy: MilestoneSortMetric, limit: Int = 5) -> [String] {
        switch sortBy {
        case .reads:
            return storyIds.sorted { a, b in
                totalReads(for: a, range: .allTime) > totalReads(for: b, range: .allTime)
            }.prefix(limit).map { $0 }
        case .credits:
            return storyIds.sorted { a, b in
                (storyAnalytics[a]?.creditsEarnedFromReads ?? 0) > (storyAnalytics[b]?.creditsEarnedFromReads ?? 0)
            }.prefix(limit).map { $0 }
        case .followers:
            return storyIds.sorted { a, b in
                (storyAnalytics[a]?.followersGainedFromStory ?? 0) > (storyAnalytics[b]?.followersGainedFromStory ?? 0)
            }.prefix(limit).map { $0 }
        }
    }
}

// MARK: - Sort Metric

enum MilestoneSortMetric: String, CaseIterable, Identifiable {
    case reads
    case credits
    case followers

    var id: String { rawValue }

    var label: String {
        switch self {
        case .reads: "By reads"
        case .credits: "By credits"
        case .followers: "By followers"
        }
    }
}
